network/src/peers/peer/handshake.rs

```diff
@@ -0,0 +1,288 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::net::SocketAddr;
+
+use snow::TransportState;
+use tokio::{
+    io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt},
+    net::TcpStream,
+};
+
+use crate::{
+    peer::{cipher::Cipher, network::PeerNetwork},
+    stats,
+    NetworkError,
+    Peer,
+    Version,
+};
+
+pub struct HandshakeData {
+    pub version: Version,
+    pub noise: TransportState,
+    pub buffer: Box<[u8]>,
+    pub noise_buffer: Box<[u8]>,
+}
+
+async fn responder_handshake<W: AsyncWrite + Unpin, R: AsyncRead + Unpin>(
+    remote_address: SocketAddr,
+    own_version: &Version,
```

Comment:
if I recall correctly, `Version` is now `Copy`, so no need to borrow it

Reply:
i see i actually made that change (to make it `Copy`). that was a mistake, it should be clone, since it may contain strings and such in the future. will update